services/director-v2/src/simcore_service_director_v2/modules/redis.py

```diff
@@ -0,0 +1,180 @@
+        redis = Redis.from_url(settings.dsn_locks)
+        lock_per_node_provider = LocksPerNodeProvider()
+        return cls(app=app, redis=redis, lock_per_node_provider=lock_per_node_provider)
+
+    @classmethod
+    def instance(cls, app: FastAPI) -> "RedisLockManager":
+        if not hasattr(app.state, "redis_lock_manager"):
+            raise ConfigurationError(
+                "RedisLockManager client is not available. Please check the configuration."
+            )
+        return app.state.redis_lock_manager
+
+    @staticmethod
+    def _get_node_slots_key(docker_node_id: DockerNodeId) -> str:
+        return f"lock_manager.{docker_node_id}.lock_slots"
+
+    @staticmethod
+    def _get_node_lock_name(
+        docker_node_id: DockerNodeId, slot_index: NonNegativeInt
+    ) -> str:
+        return f"lock_manager.{docker_node_id}.lock_slot.{slot_index}"
+
+    async def get_node_slots(self, docker_node_id: DockerNodeId) -> int:
+        """get the total amount of slots available for the node"""
+        node_slots_key = self._get_node_slots_key(docker_node_id)
+        slots: Optional[bytes] = await self.redis.get(node_slots_key)
+        if slots is not None:
+            return int(slots)
+
+        default_slots = await self.lock_per_node_provider.get(docker_node_id)
+        await self.redis.set(node_slots_key, default_slots)
+        return default_slots
+
+    @cached_property
+    def lock_extend_interval(self) -> float:
+        return self.lock_timeout / 2
+
+    async def _extend_task(self, lock: Lock) -> None:
+        while True:
+            await asyncio.sleep(self.lock_extend_interval)
+            await lock.extend(self.lock_timeout, replace_ttl=True)
+
+    async def acquire_lock(self, docker_node_id: DockerNodeId) -> Optional[Lock]:
+        """
+        Tries to acquire a lock for the provided `docker_node_id` and
+        returns one in case it succeeds.
+
+        NOTE: returns `None` if all Locks for the given `docker_node_id` are
+            already in use (locked)
+        """
+        slots = await self.get_node_slots(docker_node_id)
+        for slot in range(slots):
+            node_lock_name = self._get_node_lock_name(docker_node_id, slot)
+
+            lock = self.redis.lock(name=node_lock_name, timeout=self.lock_timeout)
+            lock_acquired = await lock.acquire(blocking=False)
+
+            if lock_acquired:
+                # In order to avoid deadlock situations, where resources are not being
+                # released, a lock with a timeout will be acquired.
+
+                # When the lock is acquired a background task which extends its
+                # validity will also be created.
+
+                # Since the lifecycle of the extend task is tied to the one of the lock
+                # the task reference is attached to the lock.
+                extend_task: Task = asyncio.create_task(self._extend_task(lock))
+                setattr(lock, EXTEND_TASK_ATTR_NAME, extend_task)
```

Comment:
Consider a mapping between logs and "extend-tasks" owned by this class instance instead of setting dynamically as an attribute of the lock object.

Reply:
That is what I precisely wanted to avoid. At this point I'll crete a class `ExtendLock` which keeps track of both the lock and extend task.

Reply:
Uhm.. yes it looks better that I first imagined.